plugin/restore/mysql/mysql.go

```diff
@@ -77,12 +85,110 @@ func (r *Restore) RestorePITR(ctx context.Context, fullBackup *bufio.Scanner, bi
+// The current mechanism is by invoking mysqlbinlog and parse the output string.
+// Maybe we should parse the raw binlog header to get better documented structure?
+// nolint
+func (r *Restore) parseBinlogEventTimestamp(ctx context.Context, binlogInfo api.BinlogInfo, binlogDir string) (int64, error) {
+	args := []string{
+		path.Join(binlogDir, binlogInfo.FileName),
+		fmt.Sprintf("--start-position %d", binlogInfo.Position),
+		// This will trick mysqlbinlog to output the binlog event header followed by a warning message telling that
+		// the --stop-position is in the middle of the binlog event.
+		// It's OK, since we are only parsing for the timestamp in the binlog event header.
+		fmt.Sprintf("--stop-position %d", binlogInfo.Position+1),
+	}
+	var buf bytes.Buffer
+	cmd := exec.CommandContext(ctx, r.mysqlbinlog.GetPath(), args...)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = &buf
+
+	if err := cmd.Run(); err != nil {
+		return 0, err
+	}
+
+	timestamp, err := parseBinlogEventTimestampImpl(buf.String())
+	if err != nil {
+		return timestamp, fmt.Errorf("failed to parse binlog event timestamp, filename[%s], position[%d], error[%w]", binlogInfo.FileName, binlogInfo.Position, err)
+	}
+
+	return timestamp, nil
+}
+
+func parseBinlogEventTimestampImpl(output string) (int64, error) {
+	lines := strings.Split(output, "\n")
+	// The mysqlbinlog output will contains a line starting with "#220421 14:49:26 server id 1",
+	// which has the timestamp we are looking for.
+	// The first occurrence is the target.
+	for _, line := range lines {
+		if strings.Contains(line, "server id") {
+			fields := strings.Fields(line)
+			// fields should starts with ["#220421", "14:49:26", "server", "id"]
+			if len(fields) < 4 ||
+				(len(fields[0]) != 7 && len(fields[1]) != 8 && fields[2] != "server" && fields[3] != "id") {
+				continue
+			}
+			date, err := time.ParseInLocation("060102 15:04:05", fmt.Sprintf("%s %s", fields[0][1:], fields[1]), time.Local)
+			if err != nil {
+				return 0, err
+			}
+			return date.Unix(), nil
+		}
+	}
+	return 0, fmt.Errorf("no timestamp found in mysqlbinlog output")
+}
+
+// Find the latest logical backup and corresponding binlog info whose time is before `targetTs`.
+// The backupList should only contain DONE backups.
+// TODO(dragonly)/TODO(zp): Use this when the apply binlog PR is ready, and remove the nolint comments.
+// nolint
+func (r *Restore) getLatestBackupBeforeOrEqualTs(ctx context.Context, backupList []*api.Backup, targetTs int64, binlogDir string) (*api.Backup, error) {
+	if len(backupList) == 0 {
+		return nil, fmt.Errorf("no valid backup")
+	}
+
+	var maxEventTsLETargetTs int64
+	var minEventTs int64 = math.MaxInt64
+	var backup *api.Backup
+	for _, b := range backupList {
+		// Parse the binlog files and convert binlog positions into MySQL server timestamps.
+		if b.Payload.BinlogInfo.FileName == "" || b.Payload.BinlogInfo.Position == 0 {
```

Comment:
Shoudl this be && instead of ||? We may have 0 as the binlog position.

Reply:
AFAIK the binlog position should start from 4, because the first 4 bytes are magic bytes \fe'bin'.

Reply:
:-) that's the theory. We're checking b.Payload.BinlogInfo == BinlogInfo{} here regardless the context of binlog.

Reply:
If invoke `mysqlbinlog` with `--start-position 0`, it will contain this warning in the output.
`mysqlbinlog: [Warning] option 'start-position': unsigned value 0 adjusted to 4.`

Reply:
Changed to checking `b.Payload.BinlogInfo` is empty.